### ISSUE TITLE
Implement Prompt-Based Reinforcement Learning with Conversion Examples

### ISSUE THREAD

**Issue by anchapin** [anchapin/ModPorter-AI#989]
## Summary

The RL module (`ai-engine/rl/`) has well-designed components — agent-specific reward modifiers, quality metrics, convergence tracking — but it has **no mechanism to actually update agent behavior**. The `RLTrainingLoop` stores episodes and calculates metrics but never modifies agent parameters or prompt strategies.

## Current State

- `rl/reward_system.py` — `RewardSignalGenerator` with agent-specific modifiers (quality_weight, correctness_emphasis, etc.)
- `rl/training_loop.py` — `RLTrainingLoop` with episodes, metrics, checkpoints
- `rl/quality_scorer.py` — `ConversionQualityScorer` for assessing output quality
- `rl/data_collection.py` — data collection infrastructure
- **None of these feed back into agent behavior**

## Proposed Solution: Prompt-Based RL (No Fine-Tuning Needed)

Instead of traditional RL weight updates, implement a **prompt optimization loop**:

1. After each conversion, score quality using the existing `ConversionQualityScorer`
2. Store **successful conversion examples** (high reward) as few-shot examples in the knowledge base / vector DB
3. On next conversion of a similar mod type, **retrieve the best prior examples via RAG**
4. Track which prompt strategies produce higher quality scores, and **adapt prompt templates** accordingly

This approach:
- Is more practical than fine-tuning at current scale
- Leverages existing RL infrastructure for scoring/tracking
- Creates a self-improving system over time
- Aligns with the B2B model where conversion quality directly impacts value

## Priority

🟡 **P1 — High** | Estimated effort: ~1 week

**Comment by anchapin** [comment]
## Research Update: GEP Pattern + Better-Fit Alternatives (Apr 18, 2026)

### What prompted this

While reviewing the [EvoMap/Evolver](https://github.com/EvoMap/evolver) library (Genome Evolution Protocol), we mapped its concepts back to what this issue proposes and found several patterns — including a Python-native alternative — that make the implementation significantly more tractable.

---

### What GEP is and how it maps here

**GEP (Genome Evolution Protocol)** is a structured framework for prompt-based agent improvement:
- **Genes** — reusable, parameterized evolution rules (think: "tighten JSON schema for entity output", "add error recovery step")
- **Capsules** — compound Genes for full conversion scenarios (e.g., a complete mob conversion sequence)
- **EvolutionEvents** — auditable log of what prompt strategy changed, when, and why
- **Strategies** — `EVOLVE_STRATEGY=harden` (reinforce known-good patterns) or `explore` (try new approaches)

This maps cleanly to PortKit:
| GEP concept | PortKit equivalent |
|-------------|-------------------|
| Gene | A reusable prompt rule for a recurring Java→Bedrock pattern (e.g., "Forge event → script event") |
| Capsule | Full conversion scenario (mob conversion sequence, recipe+model+texture bundle) |
| EvolutionEvent | Log entry: what changed in the prompt strategy, which conversions it affected, reward delta |
| `harden` strategy | Solidify high-quality conversion patterns — good fit for PortKit's B2B reliability need |

The auditable EvolutionEvents trail is genuinely valuable for a B2B product — clients want to trust the system and understand *why* it changed behavior.

---

### Why GEP as-implemented isn't the right fit for PortKit right now

1. **Node.js dependency** — PortKit is Python. A subprocess bridge or API wrapper adds friction when integrating with CrewAI.
2. **Moving to source-available** — Currently GPL-3.0 (usable), but future releases will be commercially licensed. Risky to build on before the model is clear.
3. **Prompt generator only** — Evolver outputs what the next evolution step *should be*; your pipeline still needs to pick up and apply it.

---

### Better-fit alternatives

| Pattern | Fit | Key property |
|---------|-----|--------------|
| **[DSPy](https://github.com/stanfordnlp/dspy)** (Stanford, MIT, Python) | ⭐ **Recommended** | Defines conversion as a typed `Signature(java_code: str, mod_type: str → bedrock_json: str)`, then runs a `BootstrapFewShot` or `MIPRO` optimizer over your existing quality-scored examples to find the best prompt strategy automatically |
| **[OPRO](https://arxiv.org/abs/2309.03409)** (Google DeepMind) | Good | Uses the LLM itself as the optimizer — describe the task + past scores, ask for a better prompt; no framework, but less structured |
| **[TextGrad](https://github.com/zou-group/textgrad)** (MIT, Python) | Advanced | Autodiff for LLM pipelines — treats prompt optimization like gradient descent; powerful but complex, better post-PMF |
| **[Promptbreeder](https://arxiv.org/abs/2309.16797)** (DeepMind) | Interesting | Self-referential mutation/selection of prompts; research-grade, less production-ready |

---

### Why DSPy is the strongest fit

1. **Python-native** — integrates directly with the existing CrewAI + LangChain setup, no bridge layer
2. **Typed signatures** — matches PortKit's typed conversion pipeline; `Signature` is essentially a formalized version of what this issue proposes
3. **Works with existing infrastructure** — `BootstrapFewShot` optimizer uses `ConversionQualityScorer` outputs as training signal, exactly as this issue describes
4. **RAG-compatible** — retrieving high-reward examples via the existing vector DB is a first-class DSPy pattern
5. **MIT license, actively maintained** (used in production at several companies)

**Concrete implementation sketch:**
```python
import dspy

class JavaToBedrockConversion(dspy.Signature):
    """Convert Minecraft Java mod logic to Bedrock add-on JSON."""
    java_code: str = dspy.InputField(desc="Java source or decompiled logic")
    mod_type: str = dspy.InputField(desc="entity|block|item|recipe|sound")
    bedrock_json: str = dspy.OutputField(desc="Valid Bedrock add-on JSON")

converter = dspy.ChainOfThought(JavaToBedrockConversion)

# Optimization: use high-reward conversions from ConversionQualityScorer as training examples
optimizer = dspy.BootstrapFewShot(metric=conversion_quality_metric)
optimized_converter = optimizer.compile(converter, trainset=high_reward_examples)
```

The GEP design principle worth preserving: **log every optimization step as an `EvolutionEvent`-equivalent** — what changed, why, and the quality delta. DSPy doesn't do this by default but it's easy to add a logging wrapper around `optimizer.compile()`.

---

### When to revisit

**After [#990](https://github.com/anchapin/ModPorter-AI/issues/990) and [#991](https://github.com/anchapin/ModPorter-AI/issues/991) ship LLM-powered translation.** Prompt optimization on a rule-based pipeline yields no benefit — you'd be optimizing a system that doesn't use prompts yet.

The roadmap has been updated to reflect this: see the new **AI Self-Improvement Strategy** section in `docs/ROADMAP.md`.
